ssh_proxy_server/session.py

```diff
@@ -118,14 +123,28 @@ def start(self):
         if not self._start_channels():
             return False
 
-        logging.info("session started")
+        logging.info("(%s) session started", self)
         return True
 
     def close(self):
-        if self.transport.is_active():
-            self.transport.close()
-        if self.agent:
-            self.agent.close()
+        if not self.closed:
+            self.closed = True
+            if self.agent:
+                logging.debug("(%s) session cleaning up agent ... (because paramiko IO bocks, in a new Thread)", self)
+                self.agent._close()
+                # INFO: Agent closing sequence takes 15 minutes, due to blocking IO in paramiko
+                # Paramiko agent.py tries to connect to a UNIX_SOCKET; it should be created as well (prob) BUT never is
+                # Agents starts Thread -> leads to the socket.connect blocking; only returns after .join(1000) timeout
+                threading.Thread(target=self.agent.close).start()
+                logging.debug("(%s) session agent cleaned up", self)
+            if self.transport.is_active():
+                self.transport.close()
+            logging.info("(%s) session closed", self)
+            logging.debug("(%s) closing ssh client to remote", self)
+            self.ssh_client.transport.close()
```

Comment:
File ".../session.py", line 144, in close
    self.ssh_client.transport.close()
AttributeError: 'NoneType' object has no attribute 'transport'